### PR TITLE
feat: allow specifying components when using edit-validation-sets

### PR DESCRIPTION
**Pull request by andrewphelpsj**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Hello! This should enable using `edit-validation-sets` with components. Let me know if I'm missing any testing. I don't see any tests that specifically test the allowed format of validation sets, so I don't know if that is something that we need/want to be added.

### CODE REVIEW

File: snapcraft_legacy/storeapi/v2/validation_sets.py
Comment (andrewphelpsj) [reply snapcraft_legacy/storeapi/v2/validation_sets.py:62]: Good catch, thanks!

**Comment by andrewphelpsj** [comment]
@mr-cal, for the store side, that is expected to be released this week. @suligap might have more specific details. We can hold off on merging this until that change reaches the store.

I’ll update those tests now.

**Comment by mr-cal** [comment]
> @mr-cal, for the store side, that is expected to be released this week. @suligap might have more specific details. We can hold off on merging this until that change reaches the store.
> 
> I’ll update those tests now.

Sounds good.

Once the PR lands, it will be available in the `edge` channel.  It will be part of Snapcraft 8.5, which I'm planning to cut and release to `stable` in the second half of October.  Is that timing OK for you?

**Comment by andrewphelpsj** [comment]
I think that is around the right time. I'll talk with my team to confirm timelines.

File: snapcraft_legacy/storeapi/v2/validation_sets.py
Comment (mr-cal) [inline snapcraft_legacy/storeapi/v2/validation_sets.py:83]: One more suggestion to ensure the shorthand notation gets validated correctly:

```suggestion
    components: dict[str, Literal["required", "optional", "invalid"] | Component] | None = None
```

**Review by mr-cal — CHANGES_REQUESTED** [review]

File: snapcraft_legacy/storeapi/v2/validation_sets.py
Comment (andrewphelpsj) [reply snapcraft_legacy/storeapi/v2/validation_sets.py:83]: Thanks! How would you feel about something like:
```
Presence = Literal["required", "optional", "invalid"]
```
Could be used in those thee places?

**Review by mr-cal — APPROVED** [review]
Thanks!  Once the store changes are completed and we verify it works, we can land this PR.

**Review by cmatsuoka — APPROVED** [review]
Looks good, just added a question about component revision validation.

File: snapcraft/commands/validation_sets.py
Comment (andrewphelpsj) [reply snapcraft/commands/validation_sets.py:56]: Good question, we should probably do whatever is done for snap revisions right now. I'll look.

Comment (andrewphelpsj) [reply snapcraft/commands/validation_sets.py:56]: I've confirmed with @mr-cal that this error comes from the store. It doesn't look like the store is verifying component revisions right now, I'll let them know about that. Is that something we'd consider a blocker for this PR?

Comment (mr-cal) [reply snapcraft/commands/validation_sets.py:56]: Since the validation isn't critical and components are still experimental, I think we can proceed with this PR.

**Pull request merged**
